File: android-app/app/src/main/java/com/feelbachelor/doompedia/data/repo/WikiRepository.kt
package com.feelbachelor.doompedia.data.repo

import com.feelbachelor.doompedia.data.db.ArticleWithBookmark
import com.feelbachelor.doompedia.data.db.ArticleFolderRefEntity
import com.feelbachelor.doompedia.data.db.BookmarkEntity
import com.feelbachelor.doompedia.data.db.HistoryEntity
import com.feelbachelor.doompedia.data.db.SaveFolderEntity
import com.feelbachelor.doompedia.data.db.SaveFolderSummaryRow
import com.feelbachelor.doompedia.data.db.TopicAffinityEntity
import com.feelbachelor.doompedia.data.db.WikiDao
import com.feelbachelor.doompedia.data.importer.TopicClassifier
import com.feelbachelor.doompedia.domain.ArticleCard
import com.feelbachelor.doompedia.domain.PersonalizationLevel
import com.feelbachelor.doompedia.domain.RankedCard
import com.feelbachelor.doompedia.domain.SaveFolderSummary
import com.feelbachelor.doompedia.domain.editDistanceAtMostOne
import com.feelbachelor.doompedia.domain.normalizeSearch
import com.feelbachelor.doompedia.ranking.FeedRanker
import com.feelbachelor.doompedia.ranking.RankingConfig

class WikiRepository(
    private val dao: WikiDao,
    private val rankingConfig: RankingConfig,
    private val ranker: FeedRanker,
) {
    companion object {
        const val DEFAULT_BOOKMARKS_FOLDER_ID = 1L
    }

    suspend fun ensureDefaults() {
        dao.ensureDefaultFolder(createdAt = System.currentTimeMillis())
        dao.backfillBookmarksIntoDefaultFolder()
    }

    suspend fun loadFeed(
        language: String,
        personalizationLevel: PersonalizationLevel,
        limit: Int = 50,
    ): List<RankedCard> {
        val candidates = dao.feedCandidates(language, 250).map { it.toDomain() }
        val affinity = dao.topicAffinities(language).associate { it.topicKey to it.score }
        val recentTopics = dao.recentTopics(rankingConfig.guardrails.windowSize)
        return ranker.rank(
            candidates = candidates,
            topicAffinity = affinity,
            recentlySeenTopics = recentTopics,
            level = personalizationLevel,
            limit = limit,
        )
    }

    suspend fun searchByTitle(
        language: String,
        query: String,
    ): List<ArticleCard> {
        val normalized = normalizeSearch(query)
        if (normalized.isBlank()) return emptyList()

        val maxResults = rankingConfig.search.maxResults
        val exact = dao.searchExactTitle(language, normalized, maxResults)
        val prefix = dao.searchTitlePrefix(language, normalized, maxResults)
        val alias = dao.searchAlias(language, normalized, normalized, maxResults)

        val combined = LinkedHashMap<Long, ArticleCard>()
        (exact + prefix + alias).forEach { row -> combined.putIfAbsent(row.pageId, row.toDomain()) }

        if (normalized.length >= rankingConfig.search.typoMinQueryLength) {
            val firstChar = normalized.take(1)
            val minLen = normalized.length - rankingConfig.search.typoDistance
            val maxLen = normalized.length + rankingConfig.search.typoDistance
            val typoCandidates = dao.typoCandidates(
                lang = language,
                firstChar = firstChar,
                minLen = minLen,
                maxLen = maxLen,
                limit = rankingConfig.search.maxTypoCandidates,
            )
            typoCandidates.forEach { row ->
                if (combined.containsKey(row.pageId)) return@forEach
                if (editDistanceAtMostOne(normalized, row.normalizedTitle)) {
                    combined[row.pageId] = row.toDomain()
                }
            }
        }

        return combined.values.take(maxResults)
    }

    suspend fun recordOpen(card: ArticleCard, personalizationLevel: PersonalizationLevel) {
        dao.insertHistory(
            HistoryEntity(
                pageId = card.pageId,
                topicKey = card.topicKey,
                openedAt = System.currentTimeMillis(),
            )
        )

        val levelMultiplier = rankingConfig.personalization.levels[personalizationLevel.name] ?: 0.0
        if (levelMultiplier <= 0.0) return

        val learningRate = rankingConfig.personalization.learningRates["open"] ?: 0.0
        updateTopicAffinity(
            language = card.lang,
            topicKey = card.topicKey,
            delta = learningRate * levelMultiplier,
        )
    }

    suspend fun recordLessLike(card: ArticleCard, personalizationLevel: PersonalizationLevel) {
        val levelMultiplier = rankingConfig.personalization.levels[personalizationLevel.name] ?: 0.0
        if (levelMultiplier <= 0.0) return

        val learningRate = rankingConfig.personalization.learningRates["hide"] ?: -0.5
        updateTopicAffinity(
            language = card.lang,
            topicKey = card.topicKey,
            delta = learningRate * levelMultiplier,
        )
    }

    suspend fun recordMoreLike(card: ArticleCard, personalizationLevel: PersonalizationLevel) {
        val levelMultiplier = rankingConfig.personalization.levels[personalizationLevel.name] ?: 0.0
        if (levelMultiplier <= 0.0) return

        val learningRate = rankingConfig.personalization.learningRates["like"]
            ?: rankingConfig.personalization.learningRates["bookmark"]
            ?: 0.7
        updateTopicAffinity(
            language = card.lang,
            topicKey = card.topicKey,
            delta = learningRate * levelMultiplier,
        )
    }

    suspend fun toggleBookmark(pageId: Long): Boolean {
        ensureDefaults()
        val bookmarked = dao.isBookmarked(pageId)
        if (bookmarked) {
            dao.deleteBookmark(pageId)
            dao.deleteFolderRef(DEFAULT_BOOKMARKS_FOLDER_ID, pageId)
            return false
        }
        val now = System.currentTimeMillis()
        dao.upsertBookmark(
            BookmarkEntity(
                pageId = pageId,
                createdAt = now,
            )
        )
        dao.upsertFolderRefs(
            listOf(
                ArticleFolderRefEntity(
                    folderId = DEFAULT_BOOKMARKS_FOLDER_ID,
                    pageId = pageId,
                    createdAt = now,
                ),
            )
        )
        return true
    }

    suspend fun saveFolders(): List<SaveFolderSummary> {
        ensureDefaults()
        return dao.saveFoldersWithCounts().map { it.toDomain() }
    }

    suspend fun createFolder(name: String): Boolean {
        ensureDefaults()
        val cleaned = name.trim()
        if (cleaned.isBlank()) return false
        val inserted = dao.insertFolder(
            SaveFolderEntity(
                name = cleaned,
                isDefault = false,
                createdAt = System.currentTimeMillis(),
            )
        )
        return inserted != -1L
    }

    suspend fun deleteFolder(folderId: Long): Boolean {
        ensureDefaults()
        if (folderId == DEFAULT_BOOKMARKS_FOLDER_ID) return false
        return dao.deleteFolder(folderId) > 0
    }

    suspend fun savedCards(folderId: Long, limit: Int = 400): List<ArticleCard> {
        ensureDefaults()
        return dao.savedCardsInFolder(folderId = folderId, limit = limit).map { it.toDomain() }
    }

    suspend fun selectedFolderIds(pageId: Long): Set<Long> {
        ensureDefaults()
        return dao.folderIdsForArticle(pageId).toSet()
    }

    suspend fun setFoldersForArticle(pageId: Long, folderIds: Set<Long>) {
        ensureDefaults()
        val now = System.currentTimeMillis()
        dao.clearFolderRefsForArticle(pageId)
        if (folderIds.isNotEmpty()) {
            dao.upsertFolderRefs(
                folderIds.map { folderId ->
                    ArticleFolderRefEntity(
                        folderId = folderId,
                        pageId = pageId,
                        createdAt = now,
                    )
                }
            )
        }

        if (DEFAULT_BOOKMARKS_FOLDER_ID in folderIds) {
            dao.upsertBookmark(
                BookmarkEntity(
                    pageId = pageId,
                    createdAt = now,
                )
            )
        } else {
            dao.deleteBookmark(pageId)
        }
    }

    private suspend fun updateTopicAffinity(language: String, topicKey: String, delta: Double) {
        val clamp = rankingConfig.personalization.topicClamp
        val boundedDelta = delta.coerceIn(
            -rankingConfig.personalization.dailyDriftCap,
            rankingConfig.personalization.dailyDriftCap,
        )
        val existing = dao.topicAffinity(language, topicKey)
        val newScore = ((existing?.score ?: 0.0) + boundedDelta).coerceIn(clamp.min, clamp.max)
        dao.upsertTopicAffinity(
            TopicAffinityEntity(
                lang = language,
                topicKey = topicKey,
                score = newScore,
                updatedAt = System.currentTimeMillis(),
            )
        )
    }
}

private fun SaveFolderSummaryRow.toDomain(): SaveFolderSummary {
    return SaveFolderSummary(
        folderId = folderId,
        name = name,
        isDefault = isDefault,
        articleCount = articleCount,
    )
}

private fun ArticleWithBookmark.toDomain(): ArticleCard {
    val normalizedTopic = TopicClassifier.normalizeTopic(
        rawTopic = topicKey,
        title = title,
        summary = summary,
    )
    return ArticleCard(
        pageId = pageId,
        lang = lang,
        title = title,
        normalizedTitle = normalizedTitle,
        summary = summary,
        wikiUrl = wikiUrl,
        topicKey = normalizedTopic,
        qualityScore = qualityScore,
        isDisambiguation = isDisambiguation,
        sourceRevId = sourceRevId,
        updatedAt = updatedAt,
        bookmarked = bookmarked,
    )
}

File: ios-app/Doompedia/Data/WikiRepository.swift
import Foundation

final class WikiRepository {
    private let store: SQLiteStore
    private let config: RankingConfig
    private let ranker: FeedRanker
    static let defaultBookmarksFolderID: Int64 = 1
    static let defaultReadFolderID: Int64 = 2

    init(store: SQLiteStore, config: RankingConfig) {
        self.store = store
        self.config = config
        self.ranker = FeedRanker(config: config)
    }

    func bootstrap() throws {
        try SeedBootstrapper.ensureSeedData(store: store)
        try store.ensureSaveDefaults()
    }

    func loadFeed(language: String, level: PersonalizationLevel, limit: Int = 50) throws -> [RankedCard] {
        let candidates = try store.feedCandidates(language: language, limit: 250)
        let affinities = try store.topicAffinities(language: language)
        let recentTopics = try store.recentTopics(limit: config.guardrails.windowSize)
        return ranker.rank(
            candidates: candidates,
            topicAffinity: affinities,
            recentlySeenTopics: recentTopics,
            level: level,
            limit: limit
        )
    }

    func rankCandidates(
        language: String,
        level: PersonalizationLevel,
        candidates: [ArticleCard],
        limit: Int = 50
    ) throws -> [RankedCard] {
        let affinities = try store.topicAffinities(language: language)
        let recentTopics = try store.recentTopics(limit: config.guardrails.windowSize)
        return ranker.rank(
            candidates: candidates,
            topicAffinity: affinities,
            recentlySeenTopics: recentTopics,
            level: level,
            limit: limit
        )
    }

    func search(language: String, query: String) throws -> [ArticleCard] {
        let normalized = normalizeSearch(query)
        if normalized.isEmpty { return [] }

        let maxResults = config.search.maxResults
        let exact = try store.searchExactTitle(language: language, normalizedQuery: normalized, limit: maxResults)
        let prefix = try store.searchTitlePrefix(language: language, normalizedPrefix: normalized, limit: maxResults)
        let alias = try store.searchAlias(language: language, normalizedQuery: normalized, normalizedPrefix: normalized, limit: maxResults)

        var seen: Set<Int64> = []
        var ordered: [ArticleCard] = []
        for row in (exact + prefix + alias) where seen.insert(row.pageId).inserted {
            ordered.append(row)
        }

        if normalized.count >= config.search.typoMinQueryLength,
           let firstChar = normalized.first {
            let minLen = normalized.count - config.search.typoDistance
            let maxLen = normalized.count + config.search.typoDistance
            let candidates = try store.typoCandidates(
                language: language,
                firstChar: String(firstChar),
                minLen: minLen,
                maxLen: maxLen,
                limit: config.search.maxTypoCandidates
            )

            for candidate in candidates where !seen.contains(candidate.pageId) {
                if editDistanceAtMostOne(normalized, candidate.normalizedTitle) {
                    seen.insert(candidate.pageId)
                    ordered.append(candidate)
                }
            }
        }

        return Array(ordered.prefix(maxResults))
    }

    func cacheOnlineArticles(_ rows: [SeedRow]) throws {
        guard !rows.isEmpty else { return }
        try store.upsertSeedRows(rows)
    }

    func recordOpen(card: ArticleCard, level: PersonalizationLevel) throws {
        let preferenceKeys = CardKeywords.preferenceKeys(
            title: card.title,
            summary: card.summary,
            topicKey: card.topicKey,
            maxKeys: 10
        )
        let primaryTopic = CardKeywords.primaryTopic(
            title: card.title,
            summary: card.summary,
            topicKey: card.topicKey
        )

        try store.insertHistory(pageId: card.pageId, topicKey: primaryTopic)

        let levelFactor = config.personalization.levels[level.rawValue] ?? 0.0
        guard levelFactor > 0 else { return }

        let openRate = config.personalization.learningRates["open"] ?? 0
        try updateAffinities(
            language: card.lang,
            topics: preferenceKeys,
            delta: openRate * levelFactor
        )
    }

    func recordLessLike(card: ArticleCard, level: PersonalizationLevel) throws {
        let levelFactor = config.personalization.levels[level.rawValue] ?? 0.0
        guard levelFactor > 0 else { return }

        let preferenceKeys = CardKeywords.preferenceKeys(
            title: card.title,
            summary: card.summary,
            topicKey: card.topicKey,
            maxKeys: 10
        )
        let hideRate = config.personalization.learningRates["hide"] ?? -0.5
        try updateAffinities(
            language: card.lang,
            topics: preferenceKeys,
            delta: hideRate * levelFactor
        )
    }

    func recordMoreLike(card: ArticleCard, level: PersonalizationLevel) throws {
        let levelFactor = config.personalization.levels[level.rawValue] ?? 0.0
        guard levelFactor > 0 else { return }

        let preferenceKeys = CardKeywords.preferenceKeys(
            title: card.title,
            summary: card.summary,
            topicKey: card.topicKey,
            maxKeys: 10
        )
        let likeRate = config.personalization.learningRates["like"]
            ?? config.personalization.learningRates["bookmark"]
            ?? 0.7
        try updateAffinities(
            language: card.lang,
            topics: preferenceKeys,
            delta: likeRate * levelFactor
        )
    }

    func toggleBookmark(pageId: Int64) throws -> Bool {
        try store.toggleBookmark(pageId: pageId)
    }

    func saveFolders() throws -> [SaveFolderSummary] {
        try store.saveFoldersWithCounts()
    }

    func createFolder(name: String) throws -> Bool {
        try store.createFolder(name: name)
    }

    func deleteFolder(folderID: Int64) throws -> Bool {
        try store.deleteFolder(folderID: folderID)
    }

    func selectedFolderIDs(pageID: Int64) throws -> Set<Int64> {
        try store.folderIDsForArticle(pageID: pageID)
    }

    func setFoldersForArticle(pageID: Int64, folderIDs: Set<Int64>) throws {
        try store.setFoldersForArticle(pageID: pageID, folderIDs: folderIDs)
    }

    func savedCards(folderID: Int64, language: String, readSort: ReadSort = .newestFirst) throws -> [ArticleCard] {
        try store.savedCards(folderID: folderID, language: language, readSort: readSort)
    }

    func exportFolders(selectedFolderIDs: Set<Int64>? = nil, language: String) throws -> String {
        let folders = try saveFolders()
            .filter { $0.folderId != Self.defaultReadFolderID }
            .filter { selectedFolderIDs == nil || selectedFolderIDs?.contains($0.folderId) == true }

        let payloadFolders: [[String: Any]] = try folders.map { folder in
            let cards = try savedCards(folderID: folder.folderId, language: language)
            return [
                "folderId": folder.folderId,
                "name": folder.name,
                "isDefault": folder.isDefault,
                "articlePageIds": cards.map(\.pageId),
            ]
        }

        let root: [String: Any] = [
            "version": 1,
            "format": "doompedia-folder-export",
            "exportedAt": ISO8601DateFormatter().string(from: Date()),
            "folders": payloadFolders,
        ]
        let data = try JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted, .sortedKeys])
        return String(decoding: data, as: UTF8.self)
    }

    func importFolders(payload: String) throws -> Int {
        guard
            let data = payload.data(using: .utf8),
            let root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let folders = root["folders"] as? [[String: Any]]
        else {
            throw NSError(domain: "Doompedia", code: 801, userInfo: [NSLocalizedDescriptionKey: "Invalid folder JSON"])
        }

        var linked = 0
        for folder in folders {
            guard let name = (folder["name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !name.isEmpty else { continue }

            let targetFolderID: Int64
            if name.caseInsensitiveCompare("Bookmarks") == .orderedSame {
                targetFolderID = Self.defaultBookmarksFolderID
            } else if name.caseInsensitiveCompare("Read") == .orderedSame {
                targetFolderID = Self.defaultReadFolderID
            } else {
                _ = try createFolder(name: name)
                let resolved = try saveFolders().first { $0.name.caseInsensitiveCompare(name) == .orderedSame }?.folderId
                guard let resolved else { continue }
                targetFolderID = resolved
            }

            let ids = (folder["articlePageIds"] as? [NSNumber])?.map { $0.int64Value } ?? []
            for pageID in ids {
                let existing = try selectedFolderIDs(pageID: pageID)
                let merged = existing.union([targetFolderID])
                try setFoldersForArticle(pageID: pageID, folderIDs: merged)
                linked += 1
            }
        }
        return linked
    }

    private func updateAffinity(language: String, topic: String, delta: Double) throws {
        let map = try store.topicAffinities(language: language)
        let current = map[topic] ?? 0
        let boundedDelta = max(
            -config.personalization.dailyDriftCap,
            min(config.personalization.dailyDriftCap, delta)
        )
        let minClamp = config.personalization.topicClamp.min
        let maxClamp = config.personalization.topicClamp.max
        let next = max(minClamp, min(maxClamp, current + boundedDelta))
        try store.upsertTopicAffinity(language: language, topicKey: topic, score: next)
    }

    private func updateAffinities(language: String, topics: [String], delta: Double) throws {
        let uniqueTopics = Array(
            Set(
                topics
                    .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
                    .filter { !$0.isEmpty }
            )
        ).prefix(10)
        guard !uniqueTopics.isEmpty else { return }

        let topicsArray = Array(uniqueTopics)
        let primary = topicsArray.first!
        let secondary = Array(topicsArray.dropFirst())
        let primaryDelta = delta * 0.55
        let secondaryDelta = secondary.isEmpty ? 0 : (delta * 0.45 / Double(secondary.count))

        try updateAffinity(language: language, topic: primary, delta: primaryDelta)
        for topic in secondary {
            try updateAffinity(language: language, topic: topic, delta: secondaryDelta)
        }
    }
}
